### DISCUSSION TITLE
Community overlays

### DISCUSSION THREAD

**Discussion by blelump** [the-human-colossus-foundation/oca-spec#68] [General]
It's clear the community desires enhancements to continue the growth of overlays beyond what exists in the core.

## Current ecosystem evolution / where to go next

Considerable options:
- do nothing and allow the community to find their ways in bringing `the OCA enhancements` into their use cases.
- bring basic support in the semantics layer, that is, support mechanical creation of additional overlays, but with the limitation of no support to adding the mechanics to such overlays. It means the community gets the tooling to create any overlay with proper structure (including the proper canonical form), but adding the actual mechanics (support) and extracting the value out of such overlay structure is up to the community.
- bring full support of any community overlay in any ecosystem/programming language.

The last option is unrealistic. The first option leaves no control over the [determinism mechanics](https://hackmd.io/@bYQK_qO_RLa70okz8n7TQg/ry30vwKx6) of overlays and, therefore, is open to flaws. As long as anyone thinks of an `enhancement` to the OCA core, it's still within the OCA scope. We shall, therefore, address it, which will lead us towards the second option.

2nd option is a compromise between what's doable, reasonable and pragmatic. It must address the following topics:
1. enable transferability across ecosystems of what's in the core and what's in `enhancements`
1. enforce determinism mechanics upon any community-created overlay
1. enable support for creating community-created overlays

First point is addressable via novel concept of the OCA Bundle. 2nd point is the mechanical consequence of the last point and the last point requires the most work. While addressable via OCAFile, the broad need of OCAfile support across many environments and ecosystems becomes the bottleneck. We furthermore must enable seamless overlays definition using OCAFile grammar. We, therefore, must have a meta grammar to create any overlay. Example:
```
-- parser: overlay # picks the right parser
ADD OVERLAY LABEL # the actual `type` in JSON repr
ADD SEMVER 0.0.1 # the version in `type` in JSON repr
ADD OBJECT attribute_labels 
  WITH keys attr-names # basic enforcement to have an object with keys that reflect the actual attribute names
  WITH values text 
ADD ATTR language
```

Having such meta DSL for creating overlays, we can represent all the existing core overlays and expose it for the broader audience.

Such DSL solely creates the overlay structure without touching its mechanics, i.e. it doesn't specify that the `language` in label overlay must have exactly 3-chars or that `attribute_conformance` present in conformance overlay must have exactly `M` or `O` in any `attribute_conformance` value. These are validation rules (business requirements) from the perspecitive of the overlay and while can also be imposed using OCAFiles , it can come at a later stage, if at all.

One of the examples from community is the [`separator overlay`](https://github.com/the-human-colossus-foundation/oca-spec/issues/54#issue-2172405550) and the OCAFile for it could be 
```
-- parser: overlay
ADD OVERLAY separator
ADD OBJECT dataset_separators
  WITH keys attr-names
  WITH values OBJECT
    WITH keys Text
    WITH values Text
ADD OBJECT attribute_separators
  WITH keys attr-names
  WITH values OBJECT
    WITH keys Text
    WITH values Text
```

to get the output as specified in the link.

We, therefore, must provide cross-environment tooling to extend OCAFile grammar, which produces desired overlays, and furthermore, bring an OCA Bundle that includes such overlays.

At this stage, we don't discuss the use-case-specific solutions proposed by the community but how to enable them to do what they want/need.

**Comment by carlyh-micb** [comment]
I think before this is decided there should be a discussion with the community.
There is timing and complexity to also consider. 
Can we please schedule a meeting to discuss this.

**Reply by carlyh-micb** [reply]
I also note that we will be discussing this at today's DSWG meeting and this would be a great time for anyone from the TC to present this work as well. We will be discussing alternate but I think harmonious approaches.

**Reply by blelump** [reply]
This is part of ORFG work and will eventually be (if at all) presented there. The evaluation includes a proposal for proceeding with the so-called community overlays. Nothing is decided at this stage. Feel free to share your thoughts.

**Comment by carlyh-micb** [comment]
Please see this proposal for Community contributions

https://github.com/the-human-colossus-foundation/oca-spec/discussions/70

**Comment by mitfik** [comment]
> One of the examples from community is the https://github.com/the-human-colossus-foundation/oca-spec/issues/54#issue-2172405550 and the OCAFile for it could be

I think it is really bad example as this should never be an overlay in the first place. I think we need to be careful where to draw the line between what is relevant for OCA and what is not. This separator example is simply need to support different data formats/serialization this is out of scope of OCA as mentioned under #54 and should be implemented on top of it.

but it is a very good example how the development would be chaotic if we allow anybody to mess with the stack. We won't prevent people from creating what they come up with but simple guidelines what it make sense to have as overlay and provide tooling to deal with would definitely help.

I would even suggest to skip for time being OCAFile ( it is not OCA spec) it is supporting spec which defines how to build OCA. Think of it as user interface based on that you could have different flavour of that interface (i.e. instead of english syntax you could have Japanese) the outcome would be the same. 

What we are after here is how to define core architecture which would be stable to address main goals for decentralized semantic and where to draw the line between what is `core` and what is `community`.

I think everyone agrees that lightweight `core` would be desire and the rest we can leave to the community, but the core needs to be "minimal amount" of overlays which provides possibility to capture the meaning. Without that would be hard to build distributed ecosystem and exchange data between them.

This is why in ORFg there was proposition to split it as follow:

- mechanics: which is "semantic"
- presentation: which is "how you can look at it"
- invariants: which is "business logic" 

This allows us to classify it properly from perspective of tooling and implementation but as well gives people a way to figure out where what they need should be located. Keep in mind that `presentation` and `invariants` are use cases specific and here it is purely community driven. Where `mechanics` is the essence of OCA as it is now and should be properly govern as this has impact on whole ecosystem.

**Comment by mitfik** [comment]
Moving discussion from #70 in regards of community part I would like to highlight important aspect.

Intention of OCA is to protect `structural` aspect of schema. Means capture meaning and allow for interoperability in data economy.

To achieve that we need to make sure that we can protect that part as much as possible.
`Structural bundle` should not be allowed to extended by anyone without consensus in the community means it goes through RFC process and approval Tc where the implementation of the core components is hosted.

Anything community driven goes on top of it where we can provide high level framework how to do it but without implementing the "features" for specific extension. Currently implementation supports 3 out of 4 layers which were identified:

![image](https://github.com/user-attachments/assets/88d25c77-4e36-4759-a576-26265712005a)

- Structural - which is core implementation in [oca-rs](https://github.com/THCLab/oca-rs) 
- Presentation - which is implemented in [interpres](https://github.com/THCLab/oca-interpres) library
- Transformation - which is implemented in [mmio-rs](https://github.com/THCLab/mmio-rs) 

The reason why they are split that way is due to the overall maintainability and tooling which needs to be build around it as well as responsibility segregation principle which allows to build solid ecosystem around it independently on the use cases.

**Reply by pknowl** [reply]
The logical layers represented in the above diagram make sense to me.

Note that **Framing overlays** would fit within the **Structural** category in the proposed OCA Bundle architecture. Here’s why:

- **Role in Defining Semantic Structure:** Framing overlays establish the semantic alignment of data attributes by linking them to standardized vocabularies and ontologies. This role is fundamental to the structural integrity of the data, similar to capture bases or entry codes, as it ensures that data attributes are consistently understood and interoperable across systems.

- **Foundation for Interoperability:** Structural overlays in OCA are designed to provide foundational elements that make data compatible and meaningful across different environments. Since framing overlays standardize how attributes relate to widely accepted terms (such as those in DCAT-AP or Schema.org), they contribute directly to this foundational structure, ensuring data portability and semantic coherence.

- **Essential for Data Consistency and Governance:** Like character encoding or data format, framing overlays are core to the consistent interpretation of data. They anchor the meaning of each attribute in an established vocabulary, which is critical for governance and maintaining trust within decentralized data systems.

In essence, framing overlays serve as a semantic “scaffold” that aligns closely with the Structural overlay category, supporting the primary aim of structural overlays: to enforce consistency, clarity, and interoperability across diverse data ecosystems.

**Comment by mitfik** [comment]
new Idea #92 was raised to address the problem of defining community overlays.
